# Single-Cell RNA Sequencing of Brain Metastases in Mouse Models
Here we re-analyse a data set [published](https://www.cell.com/cancer-cell/abstract/S1535-6108(24)00314-3) in the group of J. Massague.

The experimental setup contains sorted cancer cells from (brain-tropic variants of triple-negative breast cancer model MDA231 and HER2+ HCC1954), labeled tumor microenvironment (mouse cells in proximity of the cancer cells) and unlabelled mouse cells from the brain. Samples from both cell lines were pooled for sequencing and using antibody-based HTOs for demultiplexing. Experiment was performed in 2 batches.

The analysis contains the following steps:
- QC and removal of damaged cells
- sample demultiplexing using HTOs
- doublet detection
- batch correction
- cell annotation using multiple references (Allen brain atlas, ImmGen)

![Mouse cells](https://github.com/MikeKlocCZ/2025_scRNA_brain_mets_MassagueJ/blob/main/figures_examples/04_TSNE_annotation-0.png "Annotated mouse cells")

We also included cell-cycle analysis (using cyclone) and in the case of Microglia (the biggest cell population) we performed pseudotime analysis using scVelo (wrapped in a velociraptor R package)

<img src="https://github.com/MikeKlocCZ/2025_scRNA_brain_mets_MassagueJ/blob/main/figures_examples/05_TSNE_velocity.png " width="600">

The main goal of the analysis was to identify putative signals from the TME towards the cancer cells that support their metastatic collonization. To that end, we used  [NicheNet](https://nichenet.be) -- a modeling toolkit that infers intercellular communication. It contains curated signaling networks from multiple public data sources that link active **ligand-receptor** pairs with **target genes** , e. g. those that are differentially expressed upon the signalling. To run this analysis, one needs a set of potential ligands (genes expressed in a given subset of TME cells, for example Microglia subclusters), potential receptors (genes expressed by the cancer cells) and a set of differentially expressed genes in the cancer cells between the brain-tropic and parental lines (obtained from literature).  

Here is an example of the analysis for MDA231 cell line and microglia subsets,
![Cell-cell interaction analysis](https://github.com/MikeKlocCZ/2025_scRNA_brain_mets_MassagueJ/blob/main/figures_examples/05_CombinedMicroglia.png "Cell-cell interaction analysis")

## How to run the analysis
The data needed to run the analysis are in a [Zenodo](https://zenodo.org/records/16743286) repository. The content of the `*_input_data` folders needs to copied to the respective input_data folders in this github folder structure. So, `cancer_cells_input_data` should be copied to `cancer_cells/input_data` etc.

The scripts are supposed to be executed in a pre-defined order 01, 02 etc. Intermediate files are stored in `Rdata` folder that is automatically generated by the first script in each sub-directory. These intermediate `rds` files can be also directly found in the Zenodo repository in the respective `*_input_data` folders, so in principle, the scripts do not need to be run all from the beginning since every "initial file" for each script is available. However, they need to be manually copied to the `Rdata` folder for the scripts to run seamlessly.
